### DISCUSSION TITLE
Multi-rate simulation based on andes

### DISCUSSION THREAD

**Discussion by xjlcode** [CURENT/andes#435] [Ideas]
Dear Dr.Cui,
  I want to write a multi-rate simulation program in the andes framework. Do you think this idea is feasible? Or do you have any suggestions?
  Thanks.

**Comment by cuihantao** [comment]
It is definitely doable, but that sounds like a large project. What specific goals do you want to achieve?

**Reply by xjlcode** [reply]
This is just my preliminary idea.
 I would like to run two subsystems with different step sizes in andes and write interface files to connect them together.
 
 Multi-rate with different variables may be added later.

**Comment by cuihantao** [comment]
This is doable. You can create two systems objects and prepare input files accordingly. Your interface can control the time for data exchange and do the data exchange. This is known as the diakoptics, or the method of tearing.

I’m not sure about different variables using multiple rates, but I’m happy to discuss if you provide more details.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by xjlcode** [reply]
Dear Dr.Cui and Dr.Wang,
I would like to make a simple case first.
First run two subsystems, such as the IEEE14 and Kundur systems that come with the andes folder, take different steps, and then the IEEE14 system first performs a dynamic simulation, after which the voltage amplitude and phase angle of a connected node are extracted, and input them into the Kundur system, so that the Kundur system performs several more dynamic simulations, and then extracts physical quantities, and so on. Do you think this is feasible? I have extracted the voltage and phase angle of the node.

**Reply by jinningwang** [reply]
[Update2]

With guidance from Dr. Hantao Cui, one clarification should be made here:
Not being able to exchange voltage magnitudes/phase angles is the nature of variables in DAE rather than the implementation. These variables are calculated rather than input.

---

[Update]
I suddenly realized my previous answer was incorrect.
I feel this implementation is infeasible with ANDES v1.8 because it is not doable to directly alter the numeric value of the bus voltage and angle in the process of TDS.
Although you can change the initial value of bus voltage before PFlow is solved (see https://github.com/CURENT/andes/discussions/265 to check how to change the bus voltage one-time), it is not helpful for this multi-rate simulation.

---

Hello @xjlcode 

I haven't received the PhD degree, so calling me Jinning is fine. The implementation you proposed is feasible, and the workload is acceptable, to my understanding.
In this way, the entire system will be a system manufactured from a 14-bus system and a Kundur system that have one common bus, right? If so, the benchmark is also doable by simply testing the complete case file.

Regards,
Jinning

**Reply by xjlcode** [reply]
Yes，jinning，I agree with your final statement. However, I have a few questions that I don't understand.
Why is not helpful for this multi-rate simulation with changing the initial value of bus voltage before PFlow is solved？
and the trouble of phase angle, I don't quite understand.

**Reply by xjlcode** [reply]
I have encountered a problem and do not know how to break through this interface technology. 
If I directly change the voltage and phase angle of the node and run the "run" function again, the changes will be meaningless. Do I need to rewrite a new "run" function to start running from the changed voltage and phase angle?

**Reply by jinningwang** [reply]
It is because the variables are calculated from DAE.

> If I directly change the voltage and phase angle of the node and run the "run" function again, the changes will be meaningless.

As Dr. Cui commented: "It is definitely doable, but that sounds like a large project. What specific goals do you want to achieve?"

> Do I need to rewrite a new "run" function to start running from the changed voltage and phase angle?

**Reply by jinningwang** [reply]
Although you can change the initial values of voltage angles and magnitudes of the PV bus in TDS, these variables are then calculated by solving DAE.

> Yes，jinning，I agree with your final statement. However, I have a few questions that I don't understand. Why is not helpful for this multi-rate simulation with changing the initial value of bus voltage before PFlow is solved？ and the trouble of phase angle, I don't quite understand.

**Reply by xjlcode** [reply]
oh, thanks, that I understand.
Then I have no idea about multirate simulation. Do you have some Unique viewpoints？

**Comment by jinningwang** [comment]
Hello @xjlcode 

Does "two subsystems" mean separating one system into two systems in ANDES? If so, it will be two independent systems before interfacing them. In this way, I am curious about how to connect the two systems.

Regards,
Jinning

**Reply by xjlcode** [reply]
Yes, it will be two independent systems before interfacing them. 
I am trying to make interface files to connect them. It is still in test.

Regards,
xjl

**Comment by jinningwang** [comment]
This can be a reference for further development related to multi-rate simulation. Welcome further discussion if interested in such implementation.

## Overall idea
There have been some techniques to do the multi-rate simulation, such as DAE parallelism and note splitting.

## A potential proof case to multi-rate simulation
When analyzing interconnected power systems, it is sometimes possible to simplify parts of the system by using Thevenin or Norton equivalents. This can be useful for specific scenarios, although it is not a true multi-rate simulation.

The model ``PLBVFU1`` is a playback of voltage and frequency as a generator, and may be helpful for this proof case.

## DAE in ANDES
In the code, the DAE is built as an attribute ``dae`` in the class ``system`` located in file ``system.py``. DAE solving can be found in function ``run`` located in the file ``tds.py``. To be more specific, the function ``run`` calls functions ``f_update`` and ``g_update``.

For the DAE parallelism, I guess the implementation might need to restructure the DAE into several sub-DAEs. Then solve and interface them. However, this in-depth refactorization can be skipped in the proof stage.

@xjlcode Still, I would be happy to discuss potential implementation details if you clarify your desired results.

> oh, thanks, that I understand. Then I have no idea about multirate simulation. Do you have some Unique viewpoints？

**Reply by xjlcode** [reply]
Sorry, jinning. I still don't understand “Although you can change the initial values of voltage angles and magnitudes of the PV bus in TDS, these variables are then calculated by solving DAE.”

For example, I change the initial values of voltage angles and magnitudes of the PV bus of Kundur system according to the results of IEEE14 system. Then, I run the TDS for about 3 times of Kundur on the basis of the initial values of IEEE14. When it finishes, I will run the TDS of IEEE14 on the basis of the voltage angles and magnitudes of Kundur.

Is this implementation infeasible with ANDES v1.8?

**Reply by jinningwang** [reply]
This is doable **BEFORE** solving ``PFlow`` in the Kundur system.
> For example, I change the initial values of voltage angles and magnitudes of the PV bus of Kundur system according to the results of IEEE14 system.

This is doable.
> Then, I run the TDS for about 3 times of Kundur on the basis of the initial values of IEEE14.

This is doable if create a **NEW** instance of 14-bus by changing the setting points before solving ``PFlow``. I am afraid this is not doable if continue the TDS of 14-bus system, and we have discussed this as above.
> When it finishes, I will run the TDS of IEEE14 on the basis of the voltage angles and magnitudes of Kundur.

**Comment by cuihantao** [comment]
@xjlcode We would better help you if you could, more clearly, define your problem and articulate how you plan to do it. I don't think I fully understand the multi-rate you described. Finishing one simulation and use its data as an input to the other doesn't seem multi-rate. My impression of multi-rate is that you have two parallel, ongoing simulations running at different step sizes and exchanging data at defined points.

Updating voltages or other algebraic variables will not take effect because most likely the provided voltages, combines with other variables, will not satisfy the DAE. All algebraic variables are immediately replaced by the solutions from the DAE in the subsequent step. If you need to enforce exact same voltages at the boundary buses, you will need a voltage source and might look into `PLBVFU1` as Jinning proposed.
